### PR TITLE
Implement Equals for Diagnostic arguments

### PR DESCRIPTION
**Pull request by cston**
See #6262.

### CODE REVIEW

**Comment by davkean** [comment]
These look like real failures, please file a bug if they are not.

**Comment by cston** [comment]
@gafter, @AlekseyTs

**Comment by AlekseyTs** [comment]
This change definitely improves some scenarios, but the requirement for diagnostic’s arguments to implement equality remains unenforceable. Would it be better to compare textual representation of arguments instead? Then we wouldn’t rely on argument’s equality at all.

File: src/Compilers/CSharp/Portable/Symbols/SignatureOnlyParameterSymbol.cs [hunk 1]
```diff
@@ -87,5 +83,31 @@ public override bool IsImplicitlyDeclared
         internal override ModuleSymbol ContainingModule { get { throw ExceptionUtilities.Unreachable; } }
 
         #endregion Not used by MethodSignatureComparer
+
+        public override bool Equals(object obj)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/SignatureOnlyParameterSymbol.cs]: It isn't clear that this is symmetric. How do you know some other non-`SignatureOnlyParameterSymbol` won't say it is equal to a `SignatureOnlyParameterSymbol`?

**Comment by cston** [comment]
@AlekseyTs I've created a separate PR that compares string representation and changes Diagnostic.Equals to ReferenceEquals. That approach seems preferable, and I'll close this PR.

**Comment by cston** [comment]
@AlekseyTs I've created a separate PR (#6263) that compares string representation and changes Diagnostic.Equals to ReferenceEquals. That approach seems preferable, and I'll close this PR.

**Comment by cston** [comment]
Fixes #6262.

**Pull request merged**